File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataset/DQDataQuality.scala
package io.devorbit.sparklabs.batch.sales.dataset

import io.devorbit.sparklabs.batch.sales.dataframe.Utils

object DQDataQuality extends App {
  import Utils.sparkSession.implicits._

  val duplicates = DatasetUtils.customers
    .groupByKey(_.customer_id)
    .count()
    .filter(_._2 > 1)

  if (duplicates.count() == 0) println("PASS: customer_id is unique")
  else duplicates.show(false)
}

File: python/sparklabs/batch/sales/dataset/dq_data_quality.py
from sparklabs.batch.sales.dataset import utils
from sparklabs.batch.sales.dataset.models import Customer


def run() -> None:
    duplicates = utils.to_records(utils.customers, Customer)
    duplicate_count = (
        duplicates.map(lambda customer: (customer.customer_id, 1))
        .reduceByKey(lambda left, right: left + right)
        .filter(lambda pair: pair[1] > 1)
        .count()
    )
    if duplicate_count == 0:
        print("PASS: customer_id is unique")
    else:
        print(f"FAIL: found {duplicate_count} duplicate customer_id values")


if __name__ == "__main__":
    run()
